src/utils/py/README.md

```diff
@@ -1 +1,253 @@
+
+## Architecture
+
+### Layout
+
+The Python API is split into several files based on functionality:
+
+* The Python object API, located in the source tree at [src/utils/py/daos_api.py](daos_api.py).
+* The mapping of C structures to Python classes, located in the source tree at [src/utils/py/daos_cref.py](daos_cref.py)
+
+High-level abstraction classes exist to manipulate the tiers of DAOS storage:
+```python
+class DaosPool(object)
+class DaosContainer(object)
+class DaosObj(object)
+class IORequest(object)
+```
+
+`DaosPool` is a Python class representing a DAOS pool. All pool-related functionality is exposed from this class. Operations such as creating/destroying a pool, connecting to a pool, and adding a target to a storage pool are supported.
+
+`DaosContainer` is a Python class representing a DAOS container. As with the `DaosPool` class, all container-related functionality is exposed here. This class also exposes abstracted wrapper functions for the flow of creating and committing an object to a DAOS container.
+
+`DaosObj` is a Python class representing a DAOS object. Functionality such as creating/deleting objects in a container, 'punching' objects (delete an object from the specified transaction only), and object query.
+
+`IORequest` is a Python class representing a DAOS transaction. Functionality is exposed to create, delete, and modify DAOS transactions.
+
+Several classes exist for management purposes as well:
+```python
+class DaosContext(object)
+class DaosLog
+class DaosApiError(Exception)
+```
+
+`DaosContext` is a wrapper around the DAOS libraries. It is initialized with the path where DAOS libraries can be found. It describes the environmental context in which to run.
+
+`DaosLog` exposes functionality to write messages to the DAOS client log.
+
+`DaosApiError` is a custom exception class raised by the API internally in the event of a failed DAOS action.
+
+### Ctypes
+
+Ctypes is a built-in Python module for interfacing Python with existing libraries written in C/C++. The Python API is built as an object-oriented wrapper around the DAOS libraries utilizing ctypes.
+
+Ctypes documentation can be found here https://docs.python.org/3/library/ctypes.html
+
+The following demonstrates calling a C function `hello_world` from Python, with each input parameter to the C function being cast via ctypes:
+
+```C
+// hello_world.c
+
+#include <stdio.h>
+
+void hello_world(int *a, int b);
+
+void hello_world(int *a, int b) {
+    printf("hello world\n");
+}
+```
+
+```python
+# hello_world.py
+
+import ctypes
+
+# correctly cast the parameters via ctypes to match C function's expected input
+a = ctypes.byref(1)
+b = ctypes.c_int(2)
+
+# load the shared object
+my_lib = ctypes.CDLL('/full/path/to/hello_world.so')
+
+# now call it
+my_lib.hello_world(a, b)
+```
+
+In the event the C API takes in a struct as a parameter, the struct will also need to be represented via ctypes in Python:
+
+```C
+// hello_world.c
+
```

Comment:
Would be great to use a more concrete example based on the DAOS API.

Reply:
Changed to use simplified version of specific DAOS API.